src/Microsoft.AspNetCore.Routing/Internal/ParameterPolicyActivator.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Routing.Internal
+{
+    internal static class ParameterPolicyActivator
+    {
+        public static IParameterPolicy ResolveParameterPolicy(IDictionary<string, Type> inlineParameterPolicyMap, IServiceProvider serviceProvider, string inlineParameterPolicy)
+        {
+            if (inlineParameterPolicy == null)
+            {
+                throw new ArgumentNullException(nameof(inlineParameterPolicy));
+            }
+
+            string parameterPolicyKey;
+            string argumentString;
+            var indexOfFirstOpenParens = inlineParameterPolicy.IndexOf('(');
+            if (indexOfFirstOpenParens >= 0 && inlineParameterPolicy.EndsWith(")", StringComparison.Ordinal))
+            {
+                parameterPolicyKey = inlineParameterPolicy.Substring(0, indexOfFirstOpenParens);
+                argumentString = inlineParameterPolicy.Substring(
+                    indexOfFirstOpenParens + 1,
+                    inlineParameterPolicy.Length - indexOfFirstOpenParens - 2);
+            }
+            else
+            {
+                parameterPolicyKey = inlineParameterPolicy;
+                argumentString = null;
+            }
+
+            if (!inlineParameterPolicyMap.TryGetValue(parameterPolicyKey, out var parameterPolicyType))
+            {
+                throw new InvalidOperationException(Resources.FormatRoutePattern_ConstraintReferenceNotFound(
+                    parameterPolicyKey,
+                    typeof(RouteOptions),
+                    nameof(RouteOptions.ConstraintMap)));
+            }
+
+            if (!typeof(IParameterPolicy).IsAssignableFrom(parameterPolicyType))
+            {
+                throw new RouteCreationException(
+                            Resources.FormatDefaultInlineConstraintResolver_TypeNotConstraint(
+                                                        parameterPolicyType, parameterPolicyKey, typeof(IParameterPolicy).Name));
+            }
+
+            try
+            {
+                return CreateParameterPolicy(serviceProvider, parameterPolicyType, argumentString);
+            }
+            catch (RouteCreationException)
+            {
+                throw;
+            }
+            catch (Exception exception)
+            {
+                throw new RouteCreationException(
+                    $"An error occurred while trying to create an instance of '{parameterPolicyType.FullName}'.",
+                    exception);
+            }
+        }
+
+        internal static IParameterPolicy CreateParameterPolicy(IServiceProvider serviceProvider, Type constraintType, string argumentString)
```

Comment:
Rules are basic:
1. `IConvertible` params come from the inline arguments
2. Everything else is resolved from DI